protocol: Initial draft of Astoria protocol

trickeydan opened the pull request:
This is a draft specification for the Astoria Protocol, to be implemented in a reference `libastoria` as part of the kit team SR2024 roadmap. This is the first step to a stable version of Astoria, based on libastoria.

This PR merges into a `next` branch, which will be used to develop the following changes:

- MQTT v5 only. - Allows us to use the native request-response mechanism in MQTT 5. - Decouples clients from a specific topic to listen on - Allows a client to ignore the response entirely - No need for request ids - Supported by Mosquitto, MQTT.js, Paho MQTT and gmqtt.
- Renamed things - State Managers -> Services - State Consumers -> Clients - State Manager Message -> State Update Message - Broadcast Messages -> Stream Messages
- Slightly changed some message structures - Offline messages are now just blank - Changes to schema for state update message and request response - Stream messages are the same schema as broadcast messages were.
- Change `kchd` to no longer publish messages on MQTT. - We aren't doing LED control over MQTT - Let's reduce the complexity by letting kchd just consume data, rather than publish data.
- I anticipate that a service for handling user settings will end up existing - This is not yet in the spec, but the spec is designed so that a major version bump would not be required. - https://github.com/srobo/astoria/issues/187
- Usercode Service - Kill code if already running on restart.

trickeydan committed:
protocol: Initial draft of Astoria protocol

- MQTT v5 only.
    - Allows us to use the native request-response mechanism in MQTT 5.
        - Decouples clients from a specific topic to listen on
        - Allows a client to ignore the response entirely
        - No need for request ids
    - Supported by Mosquitto, MQTT.js, Paho MQTT and gmqtt.
- Renamed things
    - State Managers -> Services
    - State Consumers -> Clients
    - State Manager Message -> State Update Message
    - Broadcast Messages -> Stream Messages
- Slightly changed some message structures
    - Offline messages are now just blank
    - Changes to schema for state update message and request response
    - Stream messages are the same schema as broadcast messages were.
- Change `kchd` to no longer publish messages on MQTT.
    - We aren't doing LED control over MQTT
    - Let's reduce the complexity by letting kchd just consume data, rather than publish data.
- I anticipate that a service for handling user settings will end up existing
    - This is not yet in the spec, but the spec is designed so that a major version bump would not be required.
    - https://github.com/srobo/astoria/issues/187
- Usercode Service - Kill code if already running on restart.

WillB97 commented:
What are the messages we are using request-response for? We should probably review the use of this message structure before locking ourselves into this constraint. 

I believe setting a blank retained message actually clears the retained message from the broker, is that the intended behaviour here?

trickeydan commented:
> What are the messages we are using request-response for? We should probably review the use of this message structure before locking ourselves into this constraint.

From document:

#### Disk Service Requests

There are three possible requests to the disks service, all of which are
 primarily exist for debugging.

 -   `static_disk_add`  
     -   Payload: `{"path": "/path/to/static-disk"}`
     -   Add a new static disk.

 -   `static_disk_remove`  
     -   Payload: `{"path": "/path/to/static-disk"}`
     -   Remove an existing static disk.

 -   `static_disk_remove_all`  
     -   Payload: `{}`
     -   Remove all existing static disks.
     
#### Metadata Service Requests

 The following requests are available:

 -   `mutate`  
     -   Payload: `{"attr": "arena", "value": "A"}`
     -   Mutate a metadata attribute to a new value.
     -   The following attributes are mutable: `arena`, `zone`, `mode`

#### Usercode Service Requests

 The following requests are available:

 -   `kill`  
     -   Payload: `{}`
     -   Kill the usercode lifecycle, if there is one.

 -   `restart`<span class="title-ref"> - Payload: </span><span class="title-ref">{}</span>\`  
     -   Restart the usercode lifecycle. If code is currently running,
         kill it first.

NB: I have been considering whether we even need the disk requests, it's effectively a development tool and a different program that implements the protocol for development would probably be a better choice.

> I believe setting a blank retained message actually clears the retained message from the broker, is that the intended behaviour here?

That is the intended behaviour in the new protocol revision. We should allow services and clients to make their own decisions about how to behave when a service is unavailable, not publish a default state to the broker in the will.

RealOrangeOne commented on protocol/README.md:
```diff
@@ -0,0 +1,405 @@
+
+## Transport
+
+The protocol sends and receives messages using MQTT v5. Other versions
```
> Other versions of MQTT are not supported

Question: why?

trickeydan replied:
I've realised we've specified exactly v5, updated to v5 or later. [9d1bf66](https://github.com/srobo/astoria/pull/204/commits/9d1bf6690f1dde6b6b332bb8480c5680f91b2676)

RealOrangeOne commented on protocol/README.md:
```diff
@@ -0,0 +1,405 @@
+should still attempt to perform the requested action.
+
+A *response topic* must be either of the format
+`astoria/<slug>/response/+` or be outside of the `astoria` topic
```
> or be outside of the `astoria` topic

Suggestion: Let's not do this. Keeping things namespaced avoids any other interactions with outside systems. If we need to talk on another topic, the service can deal with that itself.

trickeydan replied:
[9d1bf66](https://github.com/srobo/astoria/pull/204/commits/9d1bf6690f1dde6b6b332bb8480c5680f91b2676)

RealOrangeOne commented on protocol/README.md:
```diff
@@ -0,0 +1,405 @@
+
+## Defined Services
+
+Every service that operates as part of Astoria should be defined in this
+specification.
```
Question: This implies that for any service to be within spec, it must be documented in the spec itself. That feels like a bad idea that won't scale. Instead, perhaps there should be either an ancillary document with each service defined, or the service should documents the details below itself, as it's only an implementation detail that these services are defined, rather than something encoded into the protocol itself.

trickeydan commented on protocol/README.md:
```diff
@@ -0,0 +1,405 @@
+# Astoria Protocol
+
+*Version 1.0.0*
```
```suggestion
*Version 1.0.0 - Draft*
```

trickeydan replied:
[9d1bf66](https://github.com/srobo/astoria/pull/204/commits/9d1bf6690f1dde6b6b332bb8480c5680f91b2676)

trickeydan committed:
Various updates to draft protocol spec

- Added DRAFT to version
- Use RFC 2119 throughout
- Specify that a service slug is snake case.
- Allow MQTT v5 or later.
- Specify that the payload MUST be UTF-8 encoded.
- Remove `service` key from messages where the service slug is in the topic.
- Clarify expected behaviour of broker when it receives an empty retained message
- Other minor edits, including typos.

trickeydan commented on protocol/README.md:
```diff
@@ -0,0 +1,405 @@
+Every service that operates as part of Astoria should be defined in this
+specification.
+
+### Disk Service
```
Added note with names of reference implementations [9d1bf66](https://github.com/srobo/astoria/pull/204/commits/9d1bf6690f1dde6b6b332bb8480c5680f91b2676)

trickeydan committed:
protocol: Use pv and v instead of long keys

We should attempt to keep the payloads to a reasonable length. As these keys are present inalmost every message, let's shorten them.

trickeydan committed:
protocol: Define the format of request encapsulation

trickeydan commented on protocol/README.md:
```diff
@@ -0,0 +1,405 @@
+`astoria/<slug>/response/+` or be outside of the `astoria` topic
+namespace to avoid collisions.
+
+A *response message* must be a JSON object of the following format:
```
46d57e4

WillB97 commented:
I've been thinking about the services that make Astoria and how they interact with the wider system. I think there are advantages to structuring the services as:
- File manager (Handles disk attach and removal, determines disk type)
  - Udiskie can execute commands on disk events which can be sent directly over MQTT
  - The service receives udiskie events over MQTT and keeps track of the attached disks (or folders)
  - On disk attach the service checks the disk to identify what types it is, reads the robot_settings.toml/astoria.json and updates the metadata, then announces the new disk and its types.
- Settings manager (handles user settings, comp mode USB, etc.)
  - Consolidates metadata messages received from file manager and clients and announces the resultant metadata.
  - This does not read any files itself
- Usercode controller
  - Runs the student's code in a subprocess and sends the logs to the logger
- Logger
  - Receives log messages from all services, including the usercode, and writes them to the log file
  - Spools messages in RAM for logging before diska is attached
- WiFi controller (coordinates both hotspot and WiFi client)
  - Controlled by metadata
- LED manager
  - Controls all the non-usercode LEDs

The clients are then:
- Web UI
- CLI
